cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,401 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#include "UrdfParser.h"
+
+#include <XmlFile.h>
+#include <XmlNode.h>
+#include <XmlParser.h>
+
+#include "Assert/Assert.h"
+
+UrdfRobotDesc UrdfParser::parse(const std::string& file_name)
+{
+    FXmlFile file;
+    file.LoadFile(UTF8_TO_TCHAR(file_name.c_str()));
+
+    FXmlNode* robot_node = file.GetRootNode();
+    ASSERT(robot_node->GetTag().Equals(TEXT("robot")));
+
+    return parseRobotNode(robot_node);
+}
+
+UrdfGeometryDesc UrdfParser::parseGeometryNode(FXmlNode* geometry_node)
+{
+    UrdfGeometryDesc geometry;
+
+    auto& child_nodes = geometry_node->GetChildrenNodes();
+    ASSERT(child_nodes.Num() == 1);
+    auto& node = child_nodes[0];
+    FString tag = node->GetTag();
+
+    if (tag.Equals(TEXT("box"))) {
+        geometry.type_ = UrdfGeometryType::Box;
+        geometry.size_ = parseVector(node->GetAttribute("size"));
+    } else if (tag.Equals(TEXT("cylinder"))) {
+        geometry.type_ = UrdfGeometryType::Cylinder;
+        geometry.length_ = FCString::Atof(*node->GetAttribute("length"));
+        geometry.radius_ = FCString::Atof(*node->GetAttribute("radius"));
+    } else if (tag.Equals(TEXT("sphere"))) {
+        geometry.type_ = UrdfGeometryType::Sphere;
+        geometry.radius_ = FCString::Atof(*node->GetAttribute("radius"));
+    } else if (tag.Equals(TEXT("mesh"))) {
+        geometry.type_ = UrdfGeometryType::Mesh;
+        geometry.filename_ = TCHAR_TO_UTF8(*node->GetAttribute("filename"));
+        geometry.scale_ = FCString::Atof(*node->GetAttribute("scale"));
+    } else {
+        // invalid geometry type
+        ASSERT(false);
+    }
+    return geometry;
+}
+
+UrdfMaterialDesc UrdfParser::parseMaterialNode(FXmlNode* material_node)
+{
+    UrdfMaterialDesc material_desc;
+
+    material_desc.name_ = TCHAR_TO_UTF8(*material_node->GetAttribute(TEXT("name")));
+
+    for (auto& node : material_node->GetChildrenNodes()) {
+        FString tag = node->GetTag();
+        if (tag.Equals(TEXT("color"))) {
+            material_desc.color_ = parseVector4(node->GetAttribute(TEXT("rgba")));
+        } else if (tag.Equals(TEXT("texture"))) {
+            material_desc.texture_ = TCHAR_TO_UTF8(*node->GetAttribute(TEXT("filename")));
+        }
+    }
+
+    return material_desc;
+}
+
+UrdfInertialDesc UrdfParser::parseInertialNode(FXmlNode* inertial_node)
+{
+    UrdfInertialDesc inertial_desc_;
+
+    bool has_origin_node = false;
```

Comment:
- `has_origin_node` is not used, so we should remove it

Reply:
fixed